File: server/src/intTest/java/org/tctalent/server/repository/db/TaskSpecificationIntTest.java
/*
 * Copyright (c) 2024 Talent Beyond Boundaries.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.tctalent.server.repository.db;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getTask;

import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.jpa.domain.Specification;
import org.tctalent.server.model.db.TaskImpl;
import org.tctalent.server.model.db.task.Task;
import org.tctalent.server.repository.db.integrationhelp.BaseDBIntegrationTest;
import org.tctalent.server.request.task.SearchTaskRequest;

public class TaskSpecificationIntTest extends BaseDBIntegrationTest {

  @Autowired
  TaskRepository repo;
  private Task task;
  Specification<TaskImpl> spec;
  SearchTaskRequest request;

  @BeforeEach
  public void setUp() {
    assertTrue(isContainerInitialised());
    repo.deleteAll();
    task = getTask(null, null);
    repo.save((TaskImpl) task);
    assertTrue(task.getId() > 0);
    assertEquals("DEFAULT", task.getName());
     request = new SearchTaskRequest();
  }

  @Test
  public void testWithKeyword() {
     request.setKeyword(task.getName());
     spec = TaskSpecification.buildSearchQuery(request);
    List<TaskImpl> result = repo.findAll(spec);
    assertNotNull(result);
    assertFalse(result.isEmpty());
    assertEquals(1, result.size());
    assertEquals(task.getName(), result.getFirst().getName());
  }

  @Test
  public void testWithNoMatchKeyword() {
    request.setKeyword("NOTHING");
    spec = TaskSpecification.buildSearchQuery(request);
    List<TaskImpl> result = repo.findAll(spec);
    assertNotNull(result);
    assertTrue(result.isEmpty());
  }
  @Test
  public void testWithEmptyKeyword() {
    request.setKeyword(null);
    spec = TaskSpecification.buildSearchQuery(request);
    List<TaskImpl> result = repo.findAll(spec);
    assertNotNull(result);
    assertFalse(result.isEmpty());
    assertEquals(1, result.size());
    assertEquals(task.getName(), result.getFirst().getName());
  }
}

File: server/src/intTest/kotlin/org/tctalent/server/repository/db/TaskSpecificationIntTest.kt
/*
 * Copyright (c) 2024 Talent Beyond Boundaries.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.tctalent.server.repository.db

import kotlin.test.*
import org.springframework.beans.factory.annotation.Autowired
import org.tctalent.server.model.db.TaskImpl
import org.tctalent.server.model.db.task.Task
import org.tctalent.server.repository.db.integrationhelp.BaseDBIntegrationTest
import org.tctalent.server.repository.db.integrationhelp.getTask
import org.tctalent.server.request.task.SearchTaskRequest

class TaskSpecificationIntTest : BaseDBIntegrationTest() {
  @Autowired lateinit var repo: TaskRepository
  private lateinit var task: Task

  @BeforeTest
  fun setup() {
    assertTrue { isContainerInitialized() }
    // Delete everything first.
    repo.deleteAll()
    task = getTask()
    repo.save(task as TaskImpl)
    assertTrue { task.id > 0 }
    assertEquals("DEFAULT", task.name)
  }

  @Test
  fun `test build search query with keyword`() {
    val request = SearchTaskRequest().apply { keyword = task.name }
    val spec = TaskSpecification.buildSearchQuery(request)
    val result = repo.findAll(spec)
    assertNotNull(result)
    assertTrue(result.isNotEmpty())
    assertEquals(1, result.size)
    assertEquals(task.id, result.first().id)
  }

  @Test
  fun `test build search query with non-matching keyword`() {
    val request = SearchTaskRequest().apply { keyword = "NOTHING" }
    val spec = TaskSpecification.buildSearchQuery(request)
    val result = repo.findAll(spec)
    assertNotNull(result)
    assertTrue(result.isEmpty())
  }

  @Test
  fun `test build search query with empty keyword`() {
    val request = SearchTaskRequest().apply { keyword = "" }
    val spec = TaskSpecification.buildSearchQuery(request)
    val result = repo.findAll(spec)
    assertNotNull(result)
    assertTrue(result.isNotEmpty())
    assertEquals(1, result.size)
    assertEquals(task.id, result.first().id)
  }
}
